### PR TITLE
Campaign settings

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This is basically Anna's PR which I could not contribute to/edit  (but I needed it merged and also the latest master pulled in to order to move forward whilst she's on vacation). The original PR is  (https://github.com/thepracticaldev/dev.to/pull/5892) and I will close it.  I added her changes and some of the fixes based on some of the suggestions.

**Anna's description:** 

I started working on settings for Shecoded 2020 campaign and further campaigns in this pr:

- added `SiteConfig` settings for campaigns:
  - `campaign_hero_html_variant_name` (name of the html variant that will be used to display campaign banner/hero over the feed)
  - `campaign_background_color`, `campaign_text_color` (these settings will be used to configure a campaign's appearance, not used in the code right now besides setting the values)
  - `campaign_featured_tags` (tags for the campaign's featured posts, not used in the code right now besides setting the values)
- added a new `campaign` group for `HtmlVariant`'s; did the related changes for the `HtmlVariant`s dashboard; improved their validation a bit

To set up a campaign hero and display it over the feed you need to:
- create an `HtmlVariant` record in `/html_variants`, choose group `campaign`. Make sure that the variant's `approved` and `published` fields are set to `true`.
- go to `SiteConfig` settings page (`/internal/config`) and set "Hero HtmlVariant name" in the campaign section to be the same that you have specified while creating an `HtmlVariant`.
- the hero html should be displayed  on the home page

To set up the campaign sidebar:
- set `sidebar_enabled` to true in the campaign section on `/internal/config` page
- set `featured_tags` on the same page
Approved posts tagged with one of the `featured_tags` should show in the sidebar.
We may need to add other settings to configure the sidebar view.
изображение

**TODO:**
-  if a user closes the full-screen display, it will not show them the display again

## Related Tickets & Documents
#5702 #5703 #5701

Rough Examples placeholders of where content would be placed:

Screenshot 2020-02-18 at 15 14 05
 
Screenshot 2020-02-18 at 15 05 56

### CODE REVIEW

**Comment by Ridhwana** [comment]
@citizen428 can you please have a look at this again, it's actually this PR https://github.com/thepracticaldev/dev.to/pull/5892 but with your suggestions added. I had to create a new PR since I wasn't able to update Anna's but needed this PR merged and up to date with master in order to move forward. I will close the other one.

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -7,7 +7,10 @@
     </div>
   <% end %>
   <div id="page-content-inner">
+    <% if @hero_html %>
+      <%= @hero_html.html_safe %>
```
Comment (citizen428) [inline app/views/layouts/application.html.erb]: Just double checking that `hero_html` is not created from user input?

Comment (Ridhwana) [reply app/views/layouts/application.html.erb]: @hero_html is from -> HtmlVariant.relevant.select(:html).find_by(group: "campaign", name: SiteConfig.campaign_hero_html_variant_name)&.html. The html for campaigns is uploaded by a very specific admin permission (not only a superuser) via internal, it's not open to the public.

Comment (citizen428) [reply app/views/layouts/application.html.erb]: Thanks, just clarifying that we're not opening ourselves up to some exploit here.

Comment (Ridhwana) [reply app/views/layouts/application.html.erb]: 💯Thanks @citizen428

**Pull request merged**